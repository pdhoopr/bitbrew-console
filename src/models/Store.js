import { types } from 'mobx-state-tree';

export default types
  .model('Store', {
    storageKey: types.optional(types.string, 'bitbrew-console'),
    token: types.maybe(types.string),
  })
  .views((self) => ({
    get isSignedIn() {
      return !!self.token;
    },
  }))
  .actions((self) => ({
    afterCreate() {
      self.token = localStorage.getItem(self.storageKey);
    },
    signIn({ accessToken }) {
      self.token = accessToken;
      localStorage.setItem(self.storageKey, accessToken);
    },
    signOut() {
      self.token = null;
      localStorage.removeItem(self.storageKey);
    },
  }));
