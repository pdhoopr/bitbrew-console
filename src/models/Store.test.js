import { getSnapshot } from 'mobx-state-tree';
import { fixtures } from '../../test';
import Store from './Store';

test('initializes the store with the default values', () => {
  const model = Store.create();
  expect(getSnapshot(model)).toMatchSnapshot();
});

test('initializes the store with the defined values', () => {
  const data = {
    storageKey: 'bitbrew:auth:token',
  };
  const model = Store.create(data);
  expect(model).toMatchObject(data);
});

test("initializes the store with a token if there's one in localStorage", () => {
  const data = {
    storageKey: 'bitbrew:auth:token',
  };
  localStorage.setItem(data.storageKey, fixtures.accessToken);
  const model = Store.create(data);
  expect(model.token).toBe(fixtures.accessToken);
});

test('signs in by setting a token in the store and localStorage', () => {
  const data = {
    accessToken: fixtures.accessToken,
  };
  const model = Store.create();
  model.signIn(data);
  expect(model.token).toBe(data.accessToken);
  expect(localStorage.getItem(model.storageKey)).toBe(data.accessToken);
  expect(model.isSignedIn).toBe(true);
});

test('signs out by removing a token from the store and localStorage', () => {
  const model = Store.create();
  model.signIn({
    accessToken: fixtures.accessToken,
  });
  model.signOut();
  expect(model.token).toBeNull();
  expect(localStorage.getItem(model.storageKey)).toBeNull();
  expect(model.isSignedIn).toBe(false);
});
