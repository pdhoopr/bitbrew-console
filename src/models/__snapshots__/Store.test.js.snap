// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initializes the store with the default values 1`] = `
Object {
  "storageKey": "bitbrew-console",
  "token": null,
}
`;
